### ISSUE TITLE
systemd unit starts vdo too early

### ISSUE THREAD

**Issue by slesru** [dm-vdo/vdo#9]
Hello!

I'm trying to get vdo work on rhel 7.4, on iscsi attached volume
 previously I tried this on centos 7.4 and local lvm volume with the same result.
Everything is OK, except startup:

 systemctl status vdo
● vdo.service - VDO volume services
   Loaded: loaded (/etc/systemd/system/vdo.service; enabled; vendor preset: enabled)
   Active: failed (Result: exit-code) since Вт 2018-03-27 20:29:01 +04; 28s ago
  Process: 2115 ExecStart=/usr/bin/vdo start --all --confFile /etc/vdoconf.yml (code=exited, status=2)
 Main PID: 2115 (code=exited, status=2)

мар 27 20:29:01 newroxar-backup systemd[1]: Starting VDO volume services...
мар 27 20:29:01 newroxar-backup vdo[2115]: vdo: ERROR - Could not set up device mapper for vdo1
мар 27 20:29:01 newroxar-backup vdo[2115]: ERROR - Could not set up device mapper for vdo1
мар 27 20:29:01 newroxar-backup vdo[2115]: vdo: ERROR - vdodumpconfig: Failed to make FileLayer from '/dev/mapper/mpathvdo' with No such file or directory
мар 27 20:29:01 newroxar-backup vdo[2115]: ERROR - vdodumpconfig: Failed to make FileLayer from '/dev/mapper/mpathvdo' with No such file or directory
мар 27 20:29:01 newroxar-backup vdo[2115]: Starting VDO vdo1
мар 27 20:29:01 newroxar-backup systemd[1]: vdo.service: main process exited, code=exited, status=2/INVALIDARGUMENT
мар 27 20:29:01 newroxar-backup systemd[1]: Failed to start VDO volume services.
мар 27 20:29:01 newroxar-backup systemd[1]: Unit vdo.service entered failed state.
мар 27 20:29:01 newroxar-backup systemd[1]: vdo.service failed.

If I start it manually after boot everything is fine.

I even tried to put it into rc.local, but no success:

мар 27 20:47:41 newroxar-backup rc.local[1929]: vdo: ERROR - Could not set up device mapper for vdo1
мар 27 20:47:41 newroxar-backup rc.local[1929]: vdo: ERROR - vdodumpconfig: Failed to make FileLayer from '/dev/mapper/mpathvdo' with No such file or directory
мар 27 20:47:41 newroxar-backup rc.local[1929]: Starting VDO vdo1
мар 27 20:47:41 newroxar-backup systemd[1]: rc-local.service: control process exited, code=exited status=2

Could you, please, add right dependency into unit ?

Thank you!

**Comment by sweettea** [comment]
Hi;

I believe your VDO systemd unit needs a dependency on /dev/mapper/mpathvdo and whatever is creating it (LVM, presumably?). I'm not sure what's making /dev/mapper/mpathvdo. If you know what service is making it or what target waits for it to exist, you could try adding a /etc/systemd/system/vdo.service.d/after-mpath.conf with contents:

```
[Service]
Requires=[whatever makes /dev/mapper/mpathvdo]
```
and see if that solves the problem.

**Comment by slesru** [comment]
Hello!

It is iscsi volume in my case, but adding Requires=iscsi.service does not help,
really I tried this and several other options before writing here.

Thank you!

**Comment by slesru** [comment]
btw, as I can see adding iscsi.service results in the same time of starting iscsi and vdo, because they are oneshot, so I added  ExecStartPre=/bin/sleep 30 ( will try to reduce time), this solved problem for me, but, anyway, this is not elegant ;-) way...

**Comment by slesru** [comment]
but, unfortunately, such change does not solve problem of correct vdo shutdown- 
if I just issue reboot vdo device is dirty on start, i.e. it was not properly shut down :-(
have no idea how to solve this...

**Comment by sweettea** [comment]
Your points are excellent that iscsi is a one-shot, making it hard to wait for all iscsi volumes to be started up before starting VDO.

 I've tried to set up a iscsi system, and it looks like the multipath stuff is a further step that I am unfamiliar with. Is there a device unit for mpathvdo? (systemctl | grep mpathvdo hopefully should show that)

Alternately, it's not universally applicable, but could you try giving VDO a Requires=remote-fs-pre.target? I'm not a master of systemd by any means, so forgive me if I'm wrong, but I think that remote-fs-pre waits for all remote devices to become available before finishing, and I'm hopeful that mpathvdo is among them. It might not be the most elegant solution, but I'm hopeful...

**Comment by cwichura** [comment]
I have run into this same problem, but with a regular local LVM volume.  System is bare metal, and has a single vg (backed by a single pv on a hardware raid controller).  Three lvs: root, swap and home.  The home lv was specified as the backing storage for the vdo container, and then an xfs filesystem created on top of vdo.

Using the x-systemd.requires= syntax in fstab WILL mount /home.  However, it results in a bunch of systemd order breakage and other services never properly start as a result.  (Problem is that vdo.service does not include DefaultDependencies=no, so requires sysinit.target, which makes for a circular reference).  I found a bug (I believe it was in the RedHat bugzilla) where it was marked "will not fix, use a unit file for the mount instead".

Problem is, the unit file doesn't work, either.  vdo.service pulls in the default dependencies, which through the chain includes the startup of LVM.  However, lvm2-monitor performs the lvm activation asynchronously, and allows system initialization to continue before it has finished all activations.   In my case, the logs generated by lvm2-monitor report the three lvs complete activation a second after vdo.service tries to start, and vdo.service complains that /dev/mapper/centos-home doesn't exist.  Attempting to start vdo.service again then works.  For whatever reason, despite being a very simple set of lvs in the vg, it's taking it about two seconds to complete the activation, with vdo.service starting about half way through that and failing.

As a workaround, I added a drop-in for vdo.service that includes an "ExecStartPre=/bin/sleep 2".  But this is inelegant.  Ideally an option should be added to vdostart that tells it to wait for the underlying storage to come available instead of just immediately bailing with a hard failure.  Then lvm activation delays, iscsi startup delays, etc., wouldn't be an issue.

**Comment by cwichura** [comment]
I suppose another solution is a generator service that is a pre-requisite to vdo.service that creates dependency files for each of the backing stores used by vdo.  E.g., vdo-pre.service that reads the vdoconfig xml and spits out something akin to:

[Unit]
Requires=dev-mapper-centos\x2dhome.device
After=dev-mapper-centos\x2dhome.device

(I've just tested that as an alternative in my vdo.service drop-in instead of the ExecStartPre=sleep and it seems to work.)

**Comment by Marcin13666** [comment]
after editing this line in /usr/lib/systemd/system/vdo.service:
`After=systemd-remount-fs.service multipathd.service multipathd.service iscsi.service iscsid.service`
vdo starts in correct way with multipathing

**Comment by slesru** [comment]
Marcin13666 , could you , please, tell me , this is on RHEL 7.6? Because on 7.4 this didn't work..

**Comment by misacek007** [comment]
I have similar setup where I have vdo on top of multipath/iscsi device `/dev/mapper/storage-big` in RHEL7.6

The default ordering of unit file will start vdo before `/dev/mapper/storage-big` is connected. This is particularily bad when system reboots and this device is not available.

`multipathd.service` says `After=iscsi.service`
`iscsi.service` says `Before=remote-fs-pre.target` 
so as a workaround I added the following override to default vdo.service unit. 
```
$ cat /etc/systemd/system/vdo.service.d/override.conf`
[Unit]
After=remote-fs-pre.target multipathd.service
```
Adding this will connect vdo device after reboot without manual intervention.
Not sure about the need of `multipathd.service` though.

**Comment by cwichura** [comment]
I still have found the cleanest way to do this is with a unit override for vdo.service that adds the required underlying block devices as Requires=/After= entries.  (See example in my prior reply here, though I have since switched to using the /dev/disk/by-id/ references now for all my systems.)  Then it doesn't matter if its iscsi or lvm or whatever bringing up the backing storage, it waits until the backing storage is really available.  This is equivalent to what .mount and .swap system units do -- they have implicit Requres=/After= on their backing store devices.

**Comment by hostalp** [comment]
Just hit the same issue after updating CentOS 7.7.1908 to the latest packages (vdo wasn't among them though). Since then I could see vdo.service consistently starting too early.

In my case the underlying device is just a local disk with a LVM on the top of it and the problem is that the vdo.service begins starting before the appropriate lvm2-pvscan@.service instance is fully started so the device isn't accessible yet. As a workaround I edited the vdo.service (`systemctl edit vdo.service`) and added `lvm2-pvscan@8:17.service` to `After` and `Requires` though I don't really like having to add such device-specific stuff in there by hand since it would work only for that system specifically.

A more generic solution that would postpone the vdo.service startup would be better. At least rework it so that it will retry a few times.

**Comment by rhawalsh** [comment]
Hello everyone.

I'd like to direct this conversation to [this BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1706154).  I have some changes that I'm testing that will make VDO work significantly better on startup/shutdown.  Please share stacks that you would like to see tested (and/or test them yourselves as the updates are posted in the BZ).

Thanks.

**Comment by rhawalsh** [comment]
I am closing this bug with a request that any comments go to [this BZ](https://bugzilla.redhat.com/show_bug.cgi?id=1706154) as mentioned above.  I think I have a solution for this and we will be working to implement that.

**Comment by Vaibhav26Thakur** [comment]
> Hello!
> 
> I'm trying to get vdo work on rhel 7.4, on iscsi attached volume
> previously I tried this on centos 7.4 and local lvm volume with the same result.
> Everything is OK, except startup:
> 
> systemctl status vdo
> ● vdo.service - VDO volume services
> Loaded: loaded (/etc/systemd/system/vdo.service; enabled; vendor preset: enabled)
> Active: failed (Result: exit-code) since Вт 2018-03-27 20:29:01 +04; 28s ago
> Process: 2115 ExecStart=/usr/bin/vdo start --all --confFile /etc/vdoconf.yml (code=exited, status=2)
> Main PID: 2115 (code=exited, status=2)
> 
> мар 27 20:29:01 newroxar-backup systemd[1]: Starting VDO volume services...
> мар 27 20:29:01 newroxar-backup vdo[2115]: vdo: ERROR - Could not set up device mapper for vdo1
> мар 27 20:29:01 newroxar-backup vdo[2115]: ERROR - Could not set up device mapper for vdo1
> мар 27 20:29:01 newroxar-backup vdo[2115]: vdo: ERROR - vdodumpconfig: Failed to make FileLayer from '/dev/mapper/mpathvdo' with No such file or directory
> мар 27 20:29:01 newroxar-backup vdo[2115]: ERROR - vdodumpconfig: Failed to make FileLayer from '/dev/mapper/mpathvdo' with No such file or directory
> мар 27 20:29:01 newroxar-backup vdo[2115]: Starting VDO vdo1
> мар 27 20:29:01 newroxar-backup systemd[1]: vdo.service: main process exited, code=exited, status=2/INVALIDARGUMENT
> мар 27 20:29:01 newroxar-backup systemd[1]: Failed to start VDO volume services.
> мар 27 20:29:01 newroxar-backup systemd[1]: Unit vdo.service entered failed state.
> мар 27 20:29:01 newroxar-backup systemd[1]: vdo.service failed.
> 
> If I start it manually after boot everything is fine.
> 
> I even tried to put it into rc.local, but no success:
> 
> мар 27 20:47:41 newroxar-backup rc.local[1929]: vdo: ERROR - Could not set up device mapper for vdo1
> мар 27 20:47:41 newroxar-backup rc.local[1929]: vdo: ERROR - vdodumpconfig: Failed to make FileLayer from '/dev/mapper/mpathvdo' with No such file or directory
> мар 27 20:47:41 newroxar-backup rc.local[1929]: Starting VDO vdo1
> мар 27 20:47:41 newroxar-backup systemd[1]: rc-local.service: control process exited, code=exited status=2
> 
> Could you, please, add right dependency into unit ?
> 
> Thank you!

Try to remove the /etc/vdoconf.yml file and then try to enable and start the service and then create the vdo partition it will definitely work.
